File: CS50's Introduction to Computer Science/Week 1 - C/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int every_other_digit(long card);
int MultiplyAndSum(int last_digit);
int number_of_digits(long card);
bool isValidAmex(long card, int numDigit);
bool isValidMastercard(long card, int numDigits);
bool isValidVisa(long card, int numDigits);

int main(void)
{
    long card = get_long("What's your credit card number? ");
    int sum_every_other_digit = every_other_digit(card);
    int numDigits = number_of_digits(card);
    bool amex = isValidAmex(card, numDigits);
    bool mcard = isValidMastercard(card, numDigits);
    bool visa = isValidVisa(card, numDigits);

    if (sum_every_other_digit % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else if (amex == true)
    {
        printf("AMEX\n");
    }
    else if (mcard == true)
    {
        printf("MASTERCARD\n");
    }
    else if (visa == true)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

bool isValidAmex(long card, int numDigits)
{
    int first_two_digits = card / pow(10, 13);
    if ((numDigits == 15) && (first_two_digits == 34 || first_two_digits == 37))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool isValidMastercard(long card, int numDigits)
{
    int first_two_digits = card / pow(10, 14);
    if ((numDigits == 16) && (first_two_digits > 50 && first_two_digits <= 55))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool isValidVisa(long card, int numDigits)
{
    if (numDigits == 13)
    {
        int first_digit = card / pow(10, 12);

        if (first_digit == 4)
        {
            return true;
        }
    }
    else if (numDigits == 16)
    {
        int first_digit = card / pow(10, 15);

        if (first_digit == 4)
        {
            return true;
        }
    }
    return false;
}

int number_of_digits(long card)
{
    int count = 0;
    while (card > 0)
    {
        count++;
        card = card / 10;
    }
    return count;
}

int every_other_digit(long card)
{
    int sum = 0;
    bool isAlternateDigit = false;
    while (card > 0)
    {
        if (isAlternateDigit == true)
        {
            int last_digit = card % 10;
            int product = MultiplyAndSum(last_digit);
            sum = sum + product;
        }
        else
        {
            int last_digit = card % 10;
            sum = sum + last_digit;
        }
        isAlternateDigit = !isAlternateDigit;
        card = card / 10;
    }
    return sum;
}

int MultiplyAndSum(int last_digit)
{
    int multiply = last_digit * 2;
    int sum = 0;

    while (multiply > 0)
    {
        int last_digit_multiply = multiply % 10;
        sum = sum + last_digit_multiply;
        multiply = multiply / 10;
    }
    return sum;
}

File: Computer Science for Artificial Intelligence - Harvard University/CS50's Introduction to Computer Science/Week 6 - Python/credit.py
from cs50 import get_int
import math

card = get_int("What's your credit card number? ")


def luhn(card):
    second_to_last_digits = []
    other_digits = []
    switch = True
    while card > 0:
        digit = card % 10
        if switch == True:
            other_digits.append(digit)
        elif switch == False:
            second_to_last_digits.append(digit)
        card /= 10
        card = math.trunc(card)
        switch = not switch

    sum_product = 0
    for i in second_to_last_digits:
        product = 2 * i
        if product >= 0 and product < 10:
            sum_product = sum_product + product
        elif product >= 10:
            sum_product = sum_product + (product % 10) + 1

    total_sum = 0
    for i in other_digits:
        total_sum = total_sum + i

    total_sum = total_sum + sum_product
    check = total_sum % 10
    return check


card_list = [int(x) for x in str(card)]
final_check = luhn(card)
length = len(str(card))

if length != 15 and length != 16 and length != 13:
    print("INVALID")

if (length == 15 and card_list[0] == 3) and (card_list[1] == 4 or card_list[1] == 7):
    if final_check == 0:
        print("AMEX")
else:
    print("INVALID")

if (
    length == 16
    and card_list[0] == 5
    and (
        card_list[1] == 1
        or card_list[1] == 2
        or card_list[1] == 3
        or card_list[1] == 4
        or card_list[1] == 5
    )
):
    if final_check == 0:
        print("MASTERCARD")
else:
    print("INVALID")

if (length == 16 or length == 13) and card_list[0] == 4:
    if final_check == 0:
        print("VISA")
else:
    print("INVALID")
